option java_package = "org.jetbrains.kni.indexer";
option java_generic_services = false;

message Function {
    required string name = 1;
    required string return_type = 2;
    repeated Parameter parameter = 3;

    message Parameter {
        required string name = 1;
        required string type = 2;
    }
}

message ObjCMethod {
    required Function function = 1;
    required bool class_method = 2;
}

message ObjCProperty {
    required string name = 1;
    required string type = 2;
    // TODO: attributes like copy/readonly/nonatomic
}

message ObjCClass {
    required string name = 1;
    optional string base_class = 2;
    repeated string protocol = 3;
    repeated string category = 4;
    repeated ObjCMethod method = 5;
    repeated ObjCProperty property = 6;
}

message ObjCProtocol {
    required string name = 1;
    repeated string base_protocol = 2;
    repeated ObjCMethod method = 3;
    repeated ObjCProperty property = 4;
}

message ObjCCategory {
    required string name = 1;
    repeated string base_protocol = 2;
    repeated ObjCMethod method = 3;
    repeated ObjCProperty property = 4;
}

message CField {
    required string name = 1;
    required string type = 2;
}

message CStruct {
    required string name = 1;
    repeated CField field = 2;
}

message Diagnostic {
    required uint32 severity = 1;
    required uint32 line = 2;
    required uint32 column = 3;
    required string message = 4;
    optional string category = 5;
}

message TranslationUnit {
    required string name = 1;
    repeated ObjCClass class = 2;
    repeated ObjCProtocol protocol = 3;
    repeated ObjCCategory category = 4;
    repeated Function function = 5;
    repeated CStruct struct = 6;
    repeated Diagnostic diagnostic = 100;
}
